server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -244,7 +235,9 @@ private List<HealthIndicatorImpact> getImpacts(
                 )
             );
         }
-        if (redNonDataNonMasterNodes.isEmpty() == false || yellowNonDataNonMasterNodes.isEmpty() == false) {
+        boolean hasUnhealthyNonMasterNonDataRoles = unhealthyRoles.stream()
+            .anyMatch(role -> role.canContainData() == false && role.equals(DiscoveryNodeRole.MASTER_ROLE) == false);
+        if (hasUnhealthyNonMasterNonDataRoles) {
```

Comment:
Can be done in a follow-up PR: 

The message now is "Some cluster functionality might be unavailable." - shall we list the node roles here? 
ie. 
"The [ml, transform, ingest] functionality might be impaired."

What do you think?

Reply:
I was planning to do it in the refactoring, but I will do it now. This is a more contained PR ;) .

Reply:
Done